Packages/Tracking Preview/PhysicsHands/Runtime/UI/Scripts/PhysicsButton.cs

```diff
@@ -0,0 +1,506 @@
+                enabled = false;
+            }
+        }
+
+        protected void OnDisable()
+        {
+            if (isPressed)
+            {
+                _unpressedThisFrame = true;
+                OnUnpress?.Invoke();
+                if (isToggleable)
+                {
+                    _isToggled = !_isToggled;
+                    UpdateToggleJoint();
+                    OnToggle?.Invoke(_isToggled);
+                }
+            }
+            if (_hovered)
+            {
+                _hovered = false;
+                OnUnhover?.Invoke();
+            }
+            if (_contact)
+            {
+                _contact = false;
+                OnUncontact?.Invoke();
+            }
+        }
+
+        protected void Start()
+        {
+            if (!_buttonElement.transform.IsChildOf(transform))
+            {
+                Debug.LogError("The assigned rigidbody is not a child of the PhysicsButton. The PhysicsButton has been disabled.", this);
+                enabled = false;
+            }
+
+            SetupButton();
+
+            _provider = FindObjectOfType<PhysicsProvider>(true);
+
+            OnPress += _OnPress.Invoke;
+            OnUnpress += _OnUnpress.Invoke;
+            OnHover += _OnHover.Invoke;
+            OnUnhover += _OnUnhover.Invoke;
+            OnContact += _OnContact.Invoke;
+            OnUncontact += _OnUncontact.Invoke;
+            OnPressedAmountChanged += _OnPressedAmountChanged.Invoke;
+            OnToggle += _OnToggle.Invoke;
+        }
+
+        protected void FixedUpdate()
+        {
+            if (_provider != null)
+            {
+                ProcessPhysicsEvents();
+            }
+        }
+
+        private void ProcessPhysicsEvents()
+        {
+            if (_provider.IsObjectHovered(_buttonElement.Rigid))
+            {
+                if (!_hovered)
+                {
+                    _hovered = true;
+                    OnHover?.Invoke();
+                }
+
+                if (_provider.GetObjectState(_buttonElement.Rigid, out var state))
+                {
+                    switch (state)
+                    {
+                        case PhysicsGraspHelper.State.Hover:
+                            _contactReleaseTime = Mathf.Clamp(--_contactReleaseTime, 0, 100);
+                            if (_contact && _contactReleaseTime == 0)
+                            {
+                                _contact = false;
+                                OnUncontact?.Invoke();
+                            }
+                            break;
+                        case PhysicsGraspHelper.State.Contact:
+                        case PhysicsGraspHelper.State.Grasp:
+                            if (!_contact)
+                            {
+                                _contact = true;
+                                OnContact?.Invoke();
```

Comment:
What's the deal with contact and the use of the contact release time? Does contact end if the hand has been in contact for a certain period?

Reply:
If use don't have a contact release time the OnContact events become very very noisy as the hand will very slightly leave the area then re-enter on the same frame. As it's coming from the overlaps of the hand we can't really modify the colliders/overlap sizes.

Reply:
OK, so it's a temporal contact hysteresis thing.